src/ssa/dynamic_objects.cpp

```diff
@@ -0,0 +1,571 @@
+    }
+  }
+}
+
+/// Finds the latest assignment to lhs before location_number and:
+///  - if the assignment rhs is a symbol, continues recursively
+///  - otherwise returns the rhs
+exprt get_malloc_size_expr(
+  const exprt &lhs,
+  unsigned location_number,
+  const goto_functiont &goto_function)
+{
+  exprt result=nil_exprt();
+  unsigned result_loc_num=0;
+  forall_goto_program_instructions(it, goto_function.body)
+  {
+    if(!it->is_assign())
+      continue;
+
+    if(lhs==it->assign_lhs())
+    {
+      result=it->assign_rhs();
+      if(result.id()==ID_typecast)
+        result=to_typecast_expr(result).op();
+      result_loc_num=it->location_number;
+    }
+
+    if(it->location_number==location_number)
+      break;
+  }
+  if(result.id()==ID_symbol)
+    return get_malloc_size_expr(result, result_loc_num, goto_function);
+
+  return result;
+}
+
+bool dynamic_objectst::get_malloc_size(
+  const goto_programt::instructiont &loc,
+  const goto_functionst::goto_functiont &current_function)
+{
+  if(loc.is_assign())
+  {
+    const code_assignt &code_assign=to_code_assign(loc.get_code());
+    if(code_assign.lhs().id()==ID_symbol)
+    {
+      // we have to propagate the malloc size
+      //   in order to get the object type
+      // TODO: this only works with inlining,
+      //       and btw, this is an ugly hack
+      std::string lhs_id=
+        id2string(to_symbol_expr(code_assign.lhs()).get_identifier());
+      if(lhs_id=="malloc::malloc_size" ||
+         lhs_id=="__builtin_alloca::alloca_size")
+      {
+        goto_functionst::goto_functiont function_copy;
+        function_copy.copy_from(current_function);
+        constant_propagator_ait const_propagator(function_copy);
+        const_propagator("", function_copy, ns);
+        malloc_size=get_malloc_size_expr(
+          loc.assign_lhs(), loc.location_number, function_copy);
+      }
+    }
+  }
+  return false;
+}
+
+void dynamic_objectst::replace_malloc_rec(
+  exprt &malloc_expr,
+  const goto_programt::instructiont &loc,
+  bool with_concrete)
+{
+  if(malloc_expr.id()==ID_side_effect &&
+     to_side_effect_expr(malloc_expr).get_statement()==ID_allocate)
+  {
+    auto object_type=get_object_type();
+    auto &dynobj=create_dynamic_object(loc, object_type, "", !loop_end);
+    exprt result_expr=dynobj.address_of(malloc_expr.type());
+
+    if(with_concrete && loop_end)
+    {
+      auto &concrete_dynobj=create_dynamic_object(
+        loc, object_type, "$co", true);
+      auto concrete_select=get_concrete_object_guard(loc, concrete_dynobj);
+      concrete_dynobj.set_alloc_guard(concrete_select);
+      db.at(&loc).front().set_alloc_guard(not_exprt(concrete_select));
+
+      result_expr=if_exprt(
+        concrete_select, concrete_dynobj.address_of(malloc_expr.type()),
+        result_expr);
+    }
+
+    malloc_expr=result_expr;
+    malloc_size=nil_exprt();
+  }
+  else
+  {
+    Forall_operands(it, malloc_expr)replace_malloc_rec(*it, loc, with_concrete);
```

Comment:
Code style here is a bit off (curly braces at else not necessary, missing newline before `replace_malloc_rec`)